add: the doc for matchLabelKeys/mismatchLabelKeys in pod (anti)affinity (#43812)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/assign-pod-node.md

Note that an empty `namespaceSelector` ({}) matches all namespaces, while a null or empty `namespaces` list and
null `namespaceSelector` matches the namespace of the Pod where the rule is defined.

#### matchLabelKeys

{{< feature-state for_k8s_version="v1.29" state="alpha" >}}

As an alpha feature, Kubernetes includes an optional `matchLabelKeys` field for Pod affinity
or anti-affinity. The field specifies keys for the labels that should  match with the incoming Pod's labels, 
when satisfying the Pod (anti)affinity.

{{< note >}}
<!-- UPDATE THIS WHEN PROMOTING TO BETA -->
The `matchLabelKeys` field is a alpha-level field and is disabled by default in
Kubernetes {{< skew currentVersion >}}.
When you want to use it, you have to enable it via the
`MatchLabelKeysInPodAffinity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
{{< /note >}}

The keys are used to look up values from the pod labels; those key-value labels are combined
(using `AND`) with the match restrictions defined using the `labelSelector` field. The combined
filtering selects the set of existing pods that will be taken into Pod (anti)affinity calculation. 

> One likely use case would be to use `matchLabelKeys` with `pod-template-hash` (set on Pods

Reviewer: ```suggestion
A common use case is to use `matchLabelKeys` with `pod-template-hash` (set on Pods
```

> managed as part of a Deployment), so that a rolling upgrade won't break affinity.

Reviewer: ```suggestion
managed as part of a Deployment, where the value is unique for each revision).
Using `pod-template-hash` in `matchLabelKeys` allows you to target the Pods that belong
to the same revision as the incoming Pod, so that a rolling upgrade won't break affinity.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: application-server
...
spec:
  template:
    affinity:
      podAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: app
              operator: In
              values:
              - database
          topologyKey: topology.kubernetes.io/zone
          # Only Pods from a given rollout are taken into consideration when calculating pod affinity.
          # If you update the Deployment, the replacement Pods follow their own affinity rules
          # (if there are any defined in the new Pod template)
          matchLabelKeys: 
          - pod-template-hash
```

#### mismatchLabelKeys

{{< feature-state for_k8s_version="v1.29" state="alpha" >}}

> As an alpha feature, Kubernetes includes an optional `mismatchLabelKeys` field for Pod affinity

Reviewer: ```suggestion
Kubernetes includes an optional `mismatchLabelKeys` field for Pod affinity
```

or anti-affinity. The field specifies keys for the labels that should **not** match with the incoming Pod's labels, 
when satisfying the Pod (anti)affinity.

> {{< note >}}

Reviewer: Move up

<!-- UPDATE THIS WHEN PROMOTING TO BETA -->
The `mismatchLabelKeys` field is a alpha-level field and is disabled by default in
Kubernetes {{< skew currentVersion >}}.
When you want to use it, you have to enable it via the
`MatchLabelKeysInPodAffinity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
{{< /note >}}

> One example use case is to ensure Pods go to the domain where only Pods from the same group are scheduled in.

Reviewer: ```suggestion
One example use case is to ensure Pods go to the topology domain (node, zone, etc) where only Pods from the same tenant or team are scheduled in.
```

> In other words, you want to avoid running Pods from two different tenants on the same node at the same time.

Reviewer: ```suggestion
In other words, you want to avoid running Pods from two different tenants on the same topology domain at the same time.
```

Note that an empty `namespaceSelector` ({}) matches all namespaces, while a null or empty `namespaces` list and
null `namespaceSelector` matches the namespace of the Pod where the rule is defined.

#### matchLabelKeys

{{< feature-state for_k8s_version="v1.29" state="alpha" >}}

> `matchLabelKeys` is an optional field which specifies the keys for the labels that should match with the incoming Pod's labels, when satisfying the Pod (anti)affinity.

Reviewer 2: ```suggestion
As an alpha feature, Kubernetes includes an optional `matchLabelKeys` field for Pod affinity
or anti-affinity. The field specifies keys for the labels that should  match with the incoming Pod's labels, 
when satisfying the Pod (anti)affinity.
```

{{< note >}}

> The `matchLabelKeys` field is a alpha-level field and disabled by default in 1.29. 

Reviewer 2: I'd write:
```suggestion
<!-- UPDATE THIS WHEN PROMOTING TO BETA -->
The `matchLabelKeys` field is a alpha-level field and is disabled by default in
Kubernetes {{< skew currentVersion >}}.
```

When you want to use it, you have to enable it via the
`MatchLabelKeysInPodAffinity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
{{< /note >}}

> The keys are used to lookup values from the pod labels, those key-value labels are ANDed 
> with labelSelector to select the group of existing pods which will be taken into Pod (anti)affinity calculation. 

Reviewer 2: ```suggestion
The keys are used to look up values from the pod labels; those key-value labels are combined
(using `AND`) with the match restrictions defined using the `labelSelector` field. The combined
filtering selects the set of existing pods that will be taken into Pod (anti)affinity calculation. 
```

> One common usecase would be to use `matchLabelKeys` with `pod-template-hash` in Deployment 
> so that the rolling upgrade won't break affinity.

Reviewer 2: ```suggestion
One likely use case would be to use `matchLabelKeys` with `pod-template-hash` (set on Pods
managed as part of a Deployment), so that a rolling upgrade won't break affinity.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: application-server
...

>   affinity:

Reviewer 3: Better add `spec:` before line 385.

    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: app
            operator: In
            values:
            - database
        topologyKey: topology.kubernetes.io/zone

>         # Only Pods of the same version are taken into consideration of PodAffinity calculation.

Reviewer 2: (nit)
```suggestion
        # Only Pods from a given rollout are taken into consideration when calculating pod affinity.
        # If you update the Deployment, the replacement Pods follow their own affinity rules
        # (if there are any defined in the new Pod template)
```

        matchLabelKeys: 
        - pod-template-hash
```

#### mismatchLabelKeys

{{< feature-state for_k8s_version="v1.29" state="alpha" >}}

> mismatchLabelKeys is an optional field 
> which specifies the keys for the labels that should NOT match with the incoming Pod's labels, 
> when satisfying the Pod (anti)affinity.

Reviewer 2: ```suggestion
As an alpha feature, Kubernetes includes an optional `mismatchLabelKeys` field for Pod affinity
or anti-affinity. The field specifies keys for the labels that should **not** match with the incoming Pod's labels, 
when satisfying the Pod (anti)affinity.
```

{{< note >}}

> The `mismatchLabelKeys` field is a alpha-level field and disabled by default in 1.29. 

Reviewer 2: I'd write:
```suggestion
<!-- UPDATE THIS WHEN PROMOTING TO BETA -->
The `mismatchLabelKeys` field is a alpha-level field and is disabled by default in
Kubernetes {{< skew currentVersion >}}.
```

When you want to use it, you have to enable it via the
`MatchLabelKeysInPodAffinity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
{{< /note >}}

> One usecase is to ensure Pods go to the domain where only Pods from the same group are scheduled in.

Reviewer 2: ```suggestion
One example use case is to ensure Pods go to the domain where only Pods from the same group are scheduled in.
In other words, you want to avoid running Pods from two different tenants on the same node at the same time.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:

>     # Assume all Pods have the label of tenant.

Reviewer 2: ```suggestion
    # Assume that all relevant Pods have a "tenant" label set
```

    tenant: tenant-a

> ...

Reviewer 2: You should also make sure to set the `.spec.template.metadata.labels` field for this Deployment, and set the `tenant` label _there_.

Author: Right.

Author: Ah wait, I didn't mean to make it Deployment. Changed it to Pod.

> affinity:

Reviewer 3: I'm afraid this indentation is incorrect.
The `affinity` property should be part of the Deployment `spec`, no?

Author: Good point, actually, it should be in a more deeper place, under spec.template.affinity.

  podAffinity:      

>     # ensures the pods of this tenant land on the same node pool
>     requiredDuringSchedulingIgnoredDuringExecution:

Reviewer 2: ```suggestion
    requiredDuringSchedulingIgnoredDuringExecution:
    # ensure that pods associated with this tenant land on the correct node pool
```

    - matchLabelKeys:
        - tenant
      topologyKey: node-pool
  podAntiAffinity:  

>     # ensures only Pods from this tenant lands on the same node pool
>     requiredDuringSchedulingIgnoredDuringExecution:

Reviewer 2: ```suggestion
    requiredDuringSchedulingIgnoredDuringExecution:
    # ensure that pods associated with this tenant can't schedule to nodes used for another tenant
```

    - mismatchLabelKeys:

>         - tenant

Reviewer 2: ```suggestion
        - tenant # whatever the value of the "tenant" label for this Pod, prevent  
                 # scheduling to nodes in any pool where any Pod from a different
                 # tenant is running.
```

      labelSelector:
        matchExpressions:
        - key: tenant

>           operator: Exists

Reviewer 2: What's this `labelSelector` for?

Author: So, it narrows down the scope of this pod anti affinity, otherwise this Pod will hate Pods from daemonsets as well, for example.

Author: > otherwise this Pod will hate Pods from daemonsets as well, for example.

Wait, will it? I have to check if `NotIn` operator matches with the Pods which doesn't have the label in the first place.

Author: Confirmed it will. So, in this usecase, this labelSelector is necessary to exclude Pods which aren't supposed to have tenant label (daemonsets)

Reviewer 2: Don't just tell the reviewers - you need to explain it (at least a little) to readers as well.

Author: done.
